svm/src/account_loader.rs

```diff
@@ -175,26 +161,17 @@ pub(crate) fn load_accounts<CB: TransactionProcessingCallback>(
 fn load_transaction_accounts<CB: TransactionProcessingCallback>(
     callbacks: &CB,
     message: &SanitizedMessage,
-    nonce: Option<&NoncePartial>,
-    fee: u64,
     error_counters: &mut TransactionErrorMetrics,
     account_overrides: Option<&AccountOverrides>,
     loaded_programs: &ProgramCacheForTxBatch,
+    unique_loaded_accounts: &mut UniqueLoadedAccounts,
 ) -> Result<LoadedTransaction> {
```

Comment:
I think we should not bind ourselves to this current return type.
`LoadedTransaction` should be the transaction state that is ready to be processed immediately. We ideally are not letting that be some stateful thing we're mutating, it should just be constructed after all checks immediately before execution, having the nonce and all up-to-date account info.

Here's how I view things:

This is called in `load_accounts`.
Ideally that function should **only** populate the `unique_loaded_accounts` map, it should not populate `LoadedTransaction`. Only errors are from program accounts not loading.

We should have a *separate* function which calculates the program indices, returning error if something fails there.

Finally we should get the `LoadedTransaction` immediately before execution of each transaction. Getting the up to date account, nonce. running checks on data size, rent, etc.



My imagined flow is:
```rust
// Load all accounts into `unique_loaded_accounts`; basic sanity checks on programs.
let initial_load_results = load_accounts(txs, &mut unique_loaded_accounts/* other args */);

// Calculate program indexes given the load results.
let program_indexes_results = calculate_program_indexes(txs, &initial_load_results);

// Loop over txs to execute
for (tx, load_result, program_index_result) in txs.iter().zip(initial_load_results).zip(program_indexes_results) {
    // do error checks
    
    let loaded_transaction = load_transaction(tx, load_result, program_index_result, /* other */);
    
    // execute
    
    if executed_successfully {
        // store all accounts back into the map
        update_unique_loaded_accounts(/*stuff*/);
    } else if executed_but_failed {
        // only update fee-payer, nonce.
        limited_update_unique_loaded_accounts(/*stuff*/);
    }
} 
```

Reply:
All suggested changes as per the pseudo code are incorporated. Just an addition that I've created `Vec<Result<LoadedTransaction, TransactionError>>` before the execution loop so that it can be utilized in the commitment. One thing that can be done to improve performance is to declare this vector as `thread_local` variable to avoid re-allocation on every entry. Let me know what are your thoughts on that?

We're currently working on updating the unit tests according to these changes.

Reply:
**Update:** 
No need for pre-allocation of `Vec<Result<LoadedTransaction, TransactionError>>` after rebasing as `LoadedTransaction` is now moved inside `ExecutedTransaction`.

Unit tests also updated and verified with the latest changes.